### PR TITLE
Add KeepAliveTimeout support to WebSocketMiddleware

### LINKED ISSUE #57072: WebSocket keep-alive timeout API
## Background and Motivation

dotnet/runtime is adding an API to control the keep-alive timeout see https://github.com/dotnet/runtime/issues/48729#issuecomment-1979542776
Our WebSocket middleware exposes options to control the websocket options being used so we would want to expose an equivalent option.

## Proposed API

```diff
namespace Microsoft.AspNetCore.Builder;

public class WebSocketOptions
{
    public TimeSpan KeepAliveInterval { get; set; }
+   public TimeSpan KeepAliveTimeout { get; set; }
}
```

## Usage Examples

```csharp
builder.AddWebSockets(o =>
{
    o.KeepAliveTimeout = TimeSpan.FromMinutes(1);
});
```

### PR DESCRIPTION
**Pull request by BrennanConroy**
Closes https://github.com/dotnet/aspnetcore/issues/57072

Feature added recently in https://github.com/dotnet/runtime/pull/105841

Replaced workaround from https://github.com/dotnet/aspnetcore/pull/48892. `ServerWebSocket` -> `AbortStream`. Solves the same issue but in a different way. This is verified by the test `WebSocket_Abort_Interrupts_Pending_ReceiveAsync`

The `AbortStream` workaround handles the background keep alive which `ServerWebSocket` did not. The old workaround only worked when someone called Abort on **our** websocket impl (`ServerWebSocket`), which can only happen in user code. With the addition of an abort call from inside the runtimes websocket impl, our `ServerWebSocket.Abort` function would never be called for keep alive timeouts.

### CODE REVIEW

**Review by amcasey — APPROVED** [review]
Can you please update the API proposal?

File: src/Middleware/WebSockets/test/UnitTests/AddWebSocketsTests.cs [hunk 1]
```diff
@@ -17,14 +17,22 @@ public void AddWebSocketsConfiguresOptions()
...
         Assert.Equal(TimeSpan.FromSeconds(1000), socketOptions.KeepAliveInterval);
+        Assert.Equal(TimeSpan.FromSeconds(1234), socketOptions.KeepAliveTimeout);
         Assert.Single(socketOptions.AllowedOrigins);
         Assert.Equal("someString", socketOptions.AllowedOrigins[0]);
     }
+
+    [Fact]
+    public void ThrowsForBadOptions()
+    {
+        Assert.Throws<ArgumentOutOfRangeException>(() => new WebSocketOptions() { KeepAliveTimeout = TimeSpan.FromMicroseconds(-1) });
```
Comment (halter73) [inline src/Middleware/WebSockets/test/UnitTests/AddWebSocketsTests.cs:36]: Can we add a test showing `Timeout.InfiniteTimeSpan` and `TimeSpan.Zero` are allowed and both disable the timeout?

File: src/Middleware/WebSockets/test/UnitTests/WebSocketMiddlewareTests.cs [hunk 2]
```diff
@@ -795,6 +795,57 @@ public async Task AcceptingWebSocketRequestDisablesTimeout()
...
+                    request.Headers.Connection.Add("Upgrade");
+                    request.Headers.Upgrade.Add(new System.Net.Http.Headers.ProductHeaderValue("websocket"));
+                    request.Headers.Add(HeaderNames.SecWebSocketVersion, "13");
+                    // SecWebSocketKey required to be 16 bytes
+                    request.Headers.Add(HeaderNames.SecWebSocketKey, Convert.ToBase64String(new byte[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16 }, Base64FormattingOptions.None));
+
+                    var response = await client.SendAsync(request, HttpCompletionOption.ResponseHeadersRead);
+
+                    var ex = await Assert.ThrowsAnyAsync<Exception>(() => tcs.Task);
+                    Assert.True(ex is ConnectionAbortedException or WebSocketException, ex.GetType().FullName);
```
Comment (halter73) [inline src/Middleware/WebSockets/test/UnitTests/WebSocketMiddlewareTests.cs:843]: I figured you'd always get a `ConnectionAbortedException` if the read is canceled by `HttpContext.Abort()`. Is it ever really a `WebSocketException`?

File: src/Middleware/WebSockets/src/WebSocketOptions.cs [hunk 3]
```diff
@@ -23,6 +25,32 @@ public WebSocketOptions()
     /// </summary>
```
Comment (halter73) [inline src/Middleware/WebSockets/src/WebSocketOptions.cs:25]: Should we update this to match `WebSocketAcceptContext.KeepAliveInterval` or vice-versa?

File: src/Http/Http.Features/src/WebSocketAcceptContext.cs [hunk 4]
```diff
@@ -11,17 +11,41 @@ namespace Microsoft.AspNetCore.Http;
...
-    /// The interval to send pong frames. This is a heart-beat that keeps the connection alive.
+    /// The interval to send keep-alive frames. This is a heart-beat that keeps the connection alive.
     /// </summary>
     public virtual TimeSpan? KeepAliveInterval { get; set; }
 
+    /// <summary>
+    /// The time to wait for a Pong frame response after sending a Ping frame. If the time is exceeded the websocket will be aborted.
+    /// </summary>
+    /// <remarks>
+    /// <c>null</c> means use the value from <c>WebSocketOptions.KeepAliveTimeout</c>.
```
[the remark this answers is not in the document]
Comment (BrennanConroy) [reply src/Http/Http.Features/src/WebSocketAcceptContext.cs:30]: Different assembly, can't reference.

**Review by halter73 — APPROVED** [review]

File: src/Middleware/WebSockets/test/UnitTests/WebSocketMiddlewareTests.cs (hunk 2 above)
Comment (BrennanConroy) [reply src/Middleware/WebSockets/test/UnitTests/WebSocketMiddlewareTests.cs:843]: Yes, if the WebSocket.ReceiveAsync call occurs after the timeout. It's a race, we could reduce the chance by changing KeepAliveTimeout to a larger value, but then the test would take longer 😄 
And would still technically be a flaky test.

File: src/Middleware/WebSockets/test/UnitTests/AddWebSocketsTests.cs (hunk 1 above)
Comment (BrennanConroy) [reply src/Middleware/WebSockets/test/UnitTests/AddWebSocketsTests.cs:36]: Idk how to test that they disable the timeout though. That's internal to the websocket impl, we'd need to use HttpClient to receive messages and then write a WebSocket frame parser to inspect the messages. And testing for a lack of Ping message from the server means just waiting for some arbitrary amount of time to check if we didn't get a message.

File: src/Http/Http.Features/src/WebSocketAcceptContext.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (amcasey) [reply src/Http/Http.Features/src/WebSocketAcceptContext.cs:30]: You can still do it using different syntax, but you lose most of the benefits.

**Pull request merged**